lnrpc/rpc.proto

```diff
@@ -1689,6 +1689,13 @@ message Hop {
     can be executed without relying on a copy of the channel graph.
     */
     string pub_key = 8 [json_name = "pub_key"];
+
+    /**
+    An optional set of key-value TLV records. This is useful within the context
+    of the SendToRoute call as it allows callers to specify arbitrary K-V pairs
+    to drop off at each hop within the onion.
+    */
+    TLVRecords tlv_records = 9 [json_name = "tlv_records"];
```

Comment:
I think I clarified this above, but if a user is crafting a _custom_ route, then they'll use the existing fields (`chan_id`, etc), this is just for them to provide any application specific tags.

Reply:
Notice that the other parsing doesn't examine these fields to include then in our in-memory representation of a hop. It's possible that the user specifies another routing critical field here, but then when they go to send, and we encode the stream, it'll be rejected for having duplicate fields.

Reply:
What I mean is this: suppose we have no graph info and we receive a `SendToRoute` call. How do we know whether we should send in the old or the new onion format?

Reply:
The assumption there is that the caller was _somehow_ able to obtain enough information to populate the arguments of the `SendToRoute` call. Just as we don't know if their fees are correct, we won't know if they're providing known or even sane values, and the error sent back in that case should convey this information.

Reply:
I understand that, but we do need to decide how to encode the onion. Legacy format or tlv. Suppose we supply only the basic fields in the `SendToRoute` call, but the node requires the new onion format (not optional). How does `lnd` know it needs to encode in the new format?

Reply:
(without having a node announcement)

Reply:
Ahh I see, in that case we'd either need to expose the field on the proto message level (to encode new or not), or just try it, then have the user figure out what's wrong.

Reply:
If you just try and it doesn't work, there still needs to be a way for the user to signal the different method. So I think either way the field needs to be exposed on the proto message.

Reply:
Adding a field to the proto to distinguish between a legacy payload and the modern one. For now, I've made the legacy the default since it'll take some time for this new feature to propagate, and we even have existing bugs in the node announcement propagation.

Reply:
Ok, sgtm. Also not to break the api.

I did a small test with A->B->C. A and C support tlv payload, B not. Ran queryroutes for a payment from A to C. It correctly showed `tlvPayload` is `false` for B. But when using this route with `SendToRoute`, it returned `UnknownNextPeer`. Logs of B show `unable to find link with destination 131843:13668356:532`, which looks like the payload was still encoded in the new format.

Reply:
Using `SendPayment` for the payment A->B->C does succeed

Reply:
Which commit were you using for this test?

Reply:
I updated the existing `SendToRoute` test with this diff and was unable to repro: 
```diff
diff --git a/lntest/itest/lnd_test.go b/lntest/itest/lnd_test.go
index f28dddf28..7d47cd7fb 100644
--- a/lntest/itest/lnd_test.go
+++ b/lntest/itest/lnd_test.go
@@ -29,6 +29,8 @@ import (
 	"github.com/btcsuite/btcutil"
 	"github.com/davecgh/go-spew/spew"
 	"github.com/go-errors/errors"
+	"github.com/golang/protobuf/jsonpb"
+	"github.com/golang/protobuf/proto"
 	"github.com/lightningnetwork/lnd"
 	"github.com/lightningnetwork/lnd/chanbackup"
 	"github.com/lightningnetwork/lnd/lnrpc"
@@ -4420,6 +4422,21 @@ func testSingleHopSendToRoute(net *lntest.NetworkHarness, t *harnessTest) {
 	closeChannelAndAssert(ctxt, t, net, net.Alice, chanPointAlice, false)
 }
 
+func printRespJSON(resp proto.Message) {
+	jsonMarshaler := &jsonpb.Marshaler{
+		EmitDefaults: true,
+		Indent:       "    ",
+	}
+
+	jsonStr, err := jsonMarshaler.MarshalToString(resp)
+	if err != nil {
+		fmt.Println("unable to decode response: ", err)
+		return
+	}
+
+	fmt.Println(jsonStr)
+}
+
 // testMultiHopSendToRoute tests that payments are properly processed
 // through a provided route. We'll create the following network topology:
 //      Alice --100k--> Bob --100k--> Carol
@@ -4431,11 +4448,26 @@ func testMultiHopSendToRoute(net *lntest.NetworkHarness, t *harnessTest) {
 	const chanAmt = btcutil.Amount(100000)
 	var networkChans []*lnrpc.ChannelPoint
 
+	// Create Carol and establish a channel from Alice. Alice is the sole
+	// funder of the channel with 100k satoshis. The network topology
+	// should look like: Alice -> Carol -> Bob.
+	carolArgs := []string{"--legacyprotocol.onion"}
+	carol, err := net.NewNode("Carol", carolArgs)
+	if err != nil {
+		t.Fatalf("unable to create new nodes: %v", err)
+	}
+	defer shutdownAndAssert(net, t, carol)
+
+	ctxt, _ := context.WithTimeout(ctxb, defaultTimeout)
+	if err := net.ConnectNodes(ctxt, net.Alice, carol); err != nil {
+		t.Fatalf("unable to connect carol to alice: %v", err)
+	}
+
 	// Open a channel with 100k satoshis between Alice and Bob with Alice
 	// being the sole funder of the channel.
-	ctxt, _ := context.WithTimeout(ctxb, channelOpenTimeout)
+	ctxt, _ = context.WithTimeout(ctxb, channelOpenTimeout)
 	chanPointAlice := openChannelAndAssert(
-		ctxt, t, net, net.Alice, net.Bob,
+		ctxt, t, net, net.Alice, carol,
 		lntest.OpenChannelParams{
 			Amt: chanAmt,
 		},
@@ -4451,27 +4483,18 @@ func testMultiHopSendToRoute(net *lntest.NetworkHarness, t *harnessTest) {
 		Index: chanPointAlice.OutputIndex,
 	}
 
-	// Create Carol and establish a channel from Bob. Bob is the sole funder
-	// of the channel with 100k satoshis. The network topology should look like:
-	// Alice -> Bob -> Carol
-	carol, err := net.NewNode("Carol", nil)
-	if err != nil {
-		t.Fatalf("unable to create new nodes: %v", err)
-	}
-	defer shutdownAndAssert(net, t, carol)
-
 	ctxt, _ = context.WithTimeout(ctxb, defaultTimeout)
 	if err := net.ConnectNodes(ctxt, carol, net.Bob); err != nil {
-		t.Fatalf("unable to connect carol to alice: %v", err)
+		t.Fatalf("unable to connect carol to bob: %v", err)
 	}
 	ctxt, _ = context.WithTimeout(ctxb, defaultTimeout)
-	err = net.SendCoins(ctxt, btcutil.SatoshiPerBitcoin, net.Bob)
+	err = net.SendCoins(ctxt, btcutil.SatoshiPerBitcoin, carol)
 	if err != nil {
 		t.Fatalf("unable to send coins to bob: %v", err)
 	}
 	ctxt, _ = context.WithTimeout(ctxb, channelOpenTimeout)
 	chanPointBob := openChannelAndAssert(
-		ctxt, t, net, net.Bob, carol,
+		ctxt, t, net, carol, net.Bob,
 		lntest.OpenChannelParams{
 			Amt: chanAmt,
 		},
@@ -4510,13 +4533,12 @@ func testMultiHopSendToRoute(net *lntest.NetworkHarness, t *harnessTest) {
 		}
 	}
 
-	// Query for routes to pay from Alice to Carol.
-	// We set FinalCltvDelta to 40 since by default QueryRoutes returns
-	// the last hop with a final cltv delta of 9 where as the default in
-	// htlcswitch is 40.
+	// Query for routes to pay from Alice to Bob.  We set FinalCltvDelta to
+	// 40 since by default QueryRoutes returns the last hop with a final
+	// cltv delta of 9 where as the default in htlcswitch is 40.
 	const paymentAmt = 1000
 	routesReq := &lnrpc.QueryRoutesRequest{
-		PubKey:         carol.PubKeyStr,
+		PubKey:         net.Bob.PubKeyStr,
 		Amt:            paymentAmt,
 		FinalCltvDelta: lnd.DefaultBitcoinTimeLockDelta,
 	}
@@ -4526,11 +4548,13 @@ func testMultiHopSendToRoute(net *lntest.NetworkHarness, t *harnessTest) {
 		t.Fatalf("unable to get route: %v", err)
 	}
 
-	// Create 5 invoices for Carol, which expect a payment from Alice for 1k
+	printRespJSON(routes)
+
+	// Create 5 invoices for Bob, which expect a payment from Alice for 1k
 	// satoshis with a different preimage each time.
 	const numPayments = 5
 	_, rHashes, _, err := createPayReqs(
-		carol, paymentAmt, numPayments,
+		net.Bob, paymentAmt, numPayments,
 	)
 	if err != nil {
 		t.Fatalf("unable to create pay reqs: %v", err)
@@ -4539,7 +4563,7 @@ func testMultiHopSendToRoute(net *lntest.NetworkHarness, t *harnessTest) {
 	// We'll wait for all parties to recognize the new channels within the
 	// network.
 	ctxt, _ = context.WithTimeout(ctxb, defaultTimeout)
-	err = carol.WaitForNetworkChannelOpen(ctxt, chanPointBob)
+	err = net.Bob.WaitForNetworkChannelOpen(ctxt, chanPointBob)
 	if err != nil {
 		t.Fatalf("bob didn't advertise his channel in time: %v", err)
 	}
@@ -4585,16 +4609,16 @@ func testMultiHopSendToRoute(net *lntest.NetworkHarness, t *harnessTest) {
 	// shifted by 5k satoshis in the direction of Carol, the sink within the
 	// payment flow generated above. The order of asserts corresponds to
 	// increasing of time is needed to embed the HTLC in commitment
-	// transaction, in channel Alice->Bob->Carol, order is Carol, Bob,
+	// transaction, in channel Alice->Carol->Bob, order is Bob, Carol,
 	// Alice.
 	const amountPaid = int64(5000)
-	assertAmountPaid(t, "Bob(local) => Carol(remote)", carol,
+	assertAmountPaid(t, "Carol(local) => Bob(remote)", net.Bob,
 		bobFundPoint, int64(0), amountPaid)
-	assertAmountPaid(t, "Bob(local) => Carol(remote)", net.Bob,
+	assertAmountPaid(t, "Carol(local) => Bob(remote)", carol,
 		bobFundPoint, amountPaid, int64(0))
-	assertAmountPaid(t, "Alice(local) => Bob(remote)", net.Bob,
+	assertAmountPaid(t, "Alice(local) => Carol(remote)", carol,
 		aliceFundPoint, int64(0), amountPaid+(baseFee*numPayments))
-	assertAmountPaid(t, "Alice(local) => Bob(remote)", net.Alice,
+	assertAmountPaid(t, "Alice(local) => Carol(remote)", net.Alice,
 		aliceFundPoint, amountPaid+(baseFee*numPayments), int64(0))
 
 	ctxt, _ = context.WithTimeout(ctxb, channelCloseTimeout)
```

It constructs Alice -> Carol -> Bob, with Carol on the legacy onion and everything is forwarded properly (just pipes QueryRoutes into SendToRoute).

Reply:
I just repro'ed it again. These are the details:
A, C: commit b49a95e0c46c0967682124576428677d1f7b46f0
B: commit 6e43974d378ab70583cbae51a907d1d58d19258f

Reply:
I start clean instances of all three and then do:
`lncli-alice queryroutes <charlie_pubkey> 10000 | lncli-alice sendtoroute --pay_hash 009347e1e0240bda04d1297906a92ce6b099a544f2edd2313f1014b994932226 -`
Expected: unknown_pay_hash, but getting unknown_next_peer with strange channel id in bob's log

Reply:
Maybe it has something to do with mapping to json in `lncli`. I noticed this coming from queryroutes. `tlvPayload` does not have the underscore.
```
"tlvPayload": true,
   "tlv_records": {
       "records": {
   }
}
```

Reply:
If I try w/ an incorrect payment hash, I get:
```
⛰   lncli --network=simnet queryroutes --final_cltv_delta=40 --dest=035b483d5f595000a33da259c31c4c5759e5601a8d69ee5b1d4f949c3fb5800dd2 --amt=2 | lncli --network=simnet sendtoroute --payment_hash=23ac69c3864b68b2257e28af8520b93c08c493349da95e48535a3805526f4654 -
{
	"payment_error": "IncorrectOrUnknownPaymentDetails(amt=2000 mSAT)",
	"payment_preimage": "",
	"payment_route": null
}
```

Here's the route: 
```
⛰   lncli --network=simnet queryroutes --final_cltv_delta=40 --dest=035b483d5f595000a33da259c31c4c5759e5601a8d69ee5b1d4f949c3fb5800dd2 --amt=2
{
    "routes": [
        {
            "total_time_lock": 1307,
            "total_fees": "1",
            "total_amt": "3",
            "hops": [
                {
                    "chan_id": "1330409069740032",
                    "chan_capacity": "10000000",
                    "amt_to_forward": "2",
                    "fee": "1",
                    "expiry": 1267,
                    "amt_to_forward_msat": "2000",
                    "fee_msat": "1000",
                    "pub_key": "02ad54c092d89c24c8b174b6f4141faa53c93f835fa856267e30d9cc199ef571c5",
                    "tlv_payload": false,
                    "tlv_records": {
                        "records": {
                        }
                    }
                },
                {
                    "chan_id": "1337006139441152",
                    "chan_capacity": "1000000",
                    "amt_to_forward": "2",
                    "fee": "0",
                    "expiry": 1267,
                    "amt_to_forward_msat": "2000",
                    "fee_msat": "0",
                    "pub_key": "035b483d5f595000a33da259c31c4c5759e5601a8d69ee5b1d4f949c3fb5800dd2",
                    "tlv_payload": true,
                    "tlv_records": {
                        "records": {
                        }
                    }
                }
            ],
            "total_fees_msat": "1000",
            "total_amt_msat": "3000"
        }
    ]
}
```